rc/tools/git.kak

```diff
@@ -326,3 +336,117 @@ define-command -params 1.. \
             ;;
     esac
 }}
+
+# Options needed by git-diff-goto-source command
+declare-option -hidden str git_diff_hunk_filename
+declare-option -hidden int git_diff_hunk_line_num_start
+declare-option -hidden int git_diff_go_to_line_num
+declare-option -hidden str git_diff_git_dir
+declare-option -hidden str git_diff_section_heading
+declare-option -hidden str git_diff_cursor_column
+
+# Works within :git diff and :git show
+define-command git-diff-goto-source \
+    -docstring 'Navigate to source by pressing the enter key in hunks when git diff is displayed. Works within :git diff and :git show' %{
+    try %{
+        set-option global git_diff_git_dir %sh{
+           git rev-parse --show-toplevel
+        }
+        # We will need this later. Need to subtract 1 because a diff has an initial column
+        # for -,+,<space>
+        set-option global git_diff_cursor_column %sh{ echo $(($kak_cursor_column-1)) }
+
+        # Allowed to press <ret> only _within_ a hunk
+        # Note: - Pressing <ret> on a @@ line (i.e. a "range line") is allowed.
+        #         This action will result in navigating to section heading (see below).
+        #       - Pressing <ret> on a +++ line is allowed. This action will result in
+        #         navigating to the first line of the file
+        execute-keys -draft 'x<a-K>^\w<ret>'
+        execute-keys -draft 'x<a-K>^$<ret>'
+        execute-keys -draft 'x<a-K>^---<ret>'
+
+        # Find the source filename for the current hunk (reverse search)
+        evaluate-commands -draft %{
+            execute-keys 'x<semicolon><a-/>^\+\+\+ b([^\n]*)<ret>'
+            set-option global git_diff_hunk_filename %reg{1}
+        }
+
+        try %{
+            # Are we on a +++ line? If so simply go to the first line of the file
+            execute-keys -draft 'x<a-k>^\+\+\+<ret>'
+            # Goto line 1 as we pressed on +++
+            edit -existing "%opt{git_diff_git_dir}%opt{git_diff_hunk_filename}" 1
+        } catch %{
+            # Find the source line at which the current hunk starts (reverse search)
+            evaluate-commands -draft %{
+                execute-keys 'x<semicolon><a-/>^@@ -\d+,\d+ \+(\d+),\d+ @@ ([^\n]*)<ret>'
```

Comment:
There is not necessarily a 'section heading',  so the space after the second `@@` should be optional. Currently this takes you to an erroneous line

Reply:
@occivink Thanks for the feedback -- I've pushed up a fresh commit. I think it should have dealt with the problem. Let me know if it fixed it!

Reply:
Not quite. If there is no section heading he line `execute-keys 'x<semicolon><a-/>^@@ -\d+,\d+ \+(\d+),\d+ @@ ([^\n]*)<ret>'` will reverse-search (which might wrap around) to the previous section heading, which is a different hunk altogether. 
We're already on a range line, so I think the search can be transformed into a `<a-x><a-k>` and then distinguish the two cases: with a section heading and without. Something like that
```
try %{
    # test @@ line
    try %{
        # test @@ line with section heading
    } catch %{
        # no section heading
    }
} catch %{
    # typical case
}
```

Reply:
@occivink Good point about the potential for a wrap around. I did a `xs` to prevent that and made related changes. I'm very sure that I've fixed the issue now -- I've pushed up another commit.

(BTW The code just needs a `try %{ }` for the section header case. We don't need a catch as we already have a default location from the range line to navigate to)